### PR TITLE
[FeatureExtractorSavingUtils] Refactor PretrainedFeatureExtractor

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR refactors the class `PreTrainedFeatureExtractor`. The following changes are done to move functionality that is shared between sequence and image feature extractors into a separate file. This should unblock the PRs of [DETR](https://github.com/huggingface/transformers/pull/9998), [VIT](https://github.com/huggingface/transformers/pull/10513), and [CLIP](https://github.com/huggingface/transformers/pull/10426)

- `PreTrainedFeatureExtractor` is renamed to `PreTrainedSequenceFeatureExtractor` because it implicitly assumed that the it will treat only sequential inputs (a.k.a sequence of float values or a sequence of float vectors). `PreTrainedFeatureExtractor` was too general
- All functionality that is shared between Image and Speech feature extractors (which IMO all relates to "saving" utilities) is moved to a `FeatureExtractorSavingUtilsMixin`
- `BatchFeature` is moved from the `feature_extraction_sequence_utils.py` to `feature_extraction_common_utils.py` to be used by the `PreTrainedImageFeatureExtractor` class as well
- The tests are refactored accordingly

The following things were assumed before applying the changes.

- In the mid-term future there will only be three modalities in HF: text, sequential features (value sequence, vector sequence), image features (2D non-sequential array)
- Models, such as ViT, DETR & CLIP will call their "preprocessor" `VITFeatureExtractor`, .... IMO, feature extractor is also a fitting name for image recognition (see: https://en.wikipedia.org/wiki/Feature_extraction) so that it is assumed that for image-text or image-only models there will be a `PreTrainedImageFeatureExtractor`, a `VITFeatureExtractor`, (and maybe a VITTokenizer & VITProcessor as well, but not necessary). For vision-text models that do require both a tokenizer and a feature extractor such as CLIP it is assumed that the classes `CLIPFeatureExtractor` and `CLIPTokenizer` are wrapped into a `CLIPProcessor` class similar to `Wav2Vec2Processor`. I think this is the most important assumption that is taken here, so we should make sure we are on the same page here @LysandreJik @sgugger @patil-suraj @NielsRogge 
- Image - Text or Image - only models won't require a `BatchImageFeature` or `BatchImage`, but can just use `BatchFeature`. From looking at the code in @NielsRogge's PR here: https://github.com/huggingface/transformers/pull/10513 this seems to be the case.

# Backwards compatibility:

The class `PreTrainedFeatureExtractor` was accessible via:

```python
from transformers import PreTrainedFeatureExtractor
```

but is now replaced by `PreTrainedSequenceFeatureExtractor`. However, since `PreTrainedFeatureExtractor` so far was only available on master, this change is OK IMO.

### CODE REVIEW

File: src/transformers/feature_extraction_common_utils.py [hunk 1]
```diff
@@ -13,48 +13,47 @@
...
 
+logger = logging.get_logger(__name__)
+
+PreTrainedFeatureExtractor = Union["PreTrainedSequenceFeatureExtractor"]  # noqa: F821
+
+
 class BatchFeature(UserDict):
     r"""
-    Holds the output of the :meth:`~transformers.PreTrainedFeatureExtractor.pad` and feature extractor specific
+    Holds the output of the :meth:`~transformers.PreTrainedSequenceFeatureExtractor.pad` and feature extractor specific
```
Comment (patrickvonplaten) [inline src/transformers/feature_extraction_common_utils.py:56]: @NielsRogge, here you can just overwrite the docstring by 

```
of the :meth: .... or the :meth:`~transformers.PreTrainedImageFeatureExtractor.pad`
```

File: src/transformers/feature_extraction_common_utils.py [hunk 2]
```diff
@@ -278,12 +251,12 @@ def from_pretrained(
...
         Returns:
-            :class:`~transformers.PreTrainedFeatureExtractor`: The feature extractor object instantiated from this
-            pretrained model.
+            :class:`~transformers.PreTrainedSequenceFeatureExtractor`: The feature extractor object instantiated from
+            this pretrained model.
 
         Examples::
 
-            # We can't instantiate directly the base class `PreTrainedFeatureExtractor` so let's show the examples on a
+            # We can't instantiate directly the base class `PreTrainedSequenceFeatureExtractor` so let's show the examples on a
```
Comment (patrickvonplaten) [inline src/transformers/feature_extraction_common_utils.py:259]: Examples for ImageFeatureExtractor saving/loading can be appended to the Examples here

File: src/transformers/feature_extraction_common_utils.py [hunk 3]
```diff
@@ -13,48 +13,47 @@
...
-
-
 if TYPE_CHECKING:
     if is_torch_available():
         import torch
 
 
+logger = logging.get_logger(__name__)
+
+PreTrainedFeatureExtractor = Union["PreTrainedSequenceFeatureExtractor"]  # noqa: F821
```
Comment (LysandreJik) [inline src/transformers/feature_extraction_common_utils.py:51]: I guess this type will be filled with the feature extractors added down the line?

File: src/transformers/feature_extraction_common_utils.py [hunk 4]
```diff
@@ -192,42 +190,19 @@ def to(self, device: Union[str, "torch.device"]) -> "BatchFeature":
...
-            try:
-                setattr(self, key, value)
-            except AttributeError as err:
-                logger.error(f"Can't set {key} with value {value} for {self}")
-                raise err
-
     @classmethod
     def from_pretrained(
         cls, pretrained_model_name_or_path: Union[str, os.PathLike], **kwargs
     ) -> "PreTrainedFeatureExtractor":
```
Comment (LysandreJik) [inline src/transformers/feature_extraction_common_utils.py:202]: If you're removing quotes, might as well do it everywhere :)

```suggestion
    ) -> PreTrainedFeatureExtractor:
```

File: src/transformers/feature_extraction_common_utils.py [hunk 5]
```diff
@@ -395,21 +367,22 @@ def get_feature_extractor_dict(
         return feature_extractor_dict, kwargs
 
     @classmethod
-    def from_dict(cls, feature_extractor_dict: Dict[str, Any], **kwargs) -> "PreTrainedFeatureExtractor":
+    def from_dict(cls, feature_extractor_dict: Dict[str, Any], **kwargs) -> PreTrainedFeatureExtractor:
         """
-        Instantiates a :class:`~transformers.PreTrainedFeatureExtractor` from a Python dictionary of parameters.
+        Instantiates a :class:`~transformers.PreTrainedSequenceFeatureExtractor` from a Python dictionary of
+        parameters.
```
Comment (LysandreJik) [inline src/transformers/feature_extraction_common_utils.py:373]: This change contradicts the type. I think it should still say it will return a `PreTrainedFeatureExtractor`, as that's a mixin that would be used for sequences and images, right?

File: src/transformers/feature_extraction_common_utils.py [hunk 6]
```diff
@@ -442,18 +415,18 @@ def to_dict(self) -> Dict[str, Any]:
...
         Args:
             json_file (:obj:`str` or :obj:`os.PathLike`):
                 Path to the JSON file containing the parameters.
 
         Returns:
-            :class:`~transformers.PreTrainedFeatureExtractor`: The feature_extractor object instantiated from that JSON
-            file.
-
+            :class:`~transformers.PreTrainedSequenceFeatureExtractor`: The feature_extractor object instantiated from
+            that JSON file.
```
Comment (LysandreJik) [inline src/transformers/feature_extraction_common_utils.py:429]: Same here, the change contradicts the type.

**Review by LysandreJik — APPROVED** [review]
Looks good! I like the API. Something I would change to streamline it with our tokenizers is to have the `pad` and `_pad` methods defined in the superclass, but raise a `NotImplementedError` if they're not implemented. Similarly to `tokenize()` in the tokenization utils base.

Also, here you imported `PaddingStrategy` which is great so as to not duplicate existing objects, how would we manage this for vision models? Since the padding/cropping strategies would probably be different (i.e., `largest` instead of `longest`, as @NielsRogge was mentioning)

File: src/transformers/feature_extraction_common_utils.py [hunk 7]
```diff
@@ -193,41 +191,18 @@ def to(self, device: Union[str, "torch.device"]) -> "BatchFeature":
...
-                logger.error(f"Can't set {key} with value {value} for {self}")
-                raise err
-
     @classmethod
     def from_pretrained(
         cls, pretrained_model_name_or_path: Union[str, os.PathLike], **kwargs
     ) -> "PreTrainedFeatureExtractor":
         r"""
-        Instantiate a :class:`~transformers.PreTrainedFeatureExtractor` (or a derived class) from a pretrained feature
+        Instantiate a :class:`~transformers.SequenceFeatureExtractor` (or a derived class) from a pretrained feature
```
Comment (sgugger) [inline src/transformers/feature_extraction_common_utils.py:205]: The docstring should not be limited to the `SequenceFeatureExtractor`. I think it should just say: " Instantiates a type of he Mixin from a pretrained feature extractor" or something like that.

File: src/transformers/feature_extraction_common_utils.py [hunk 8]
```diff
@@ -238,7 +213,7 @@ def from_pretrained(
                   huggingface.co. Valid model ids can be located at the root-level, like ``bert-base-uncased``, or
                   namespaced under a user or organization name, like ``dbmdz/bert-base-german-cased``.
                 - a path to a `directory` containing a feature extractor file saved using the
-                  :func:`~transformers.PreTrainedFeatureExtractor.save_pretrained` method, e.g.,
+                  :func:`~transformers.SequenceFeatureExtractor.save_pretrained` method, e.g.,
```
Comment (sgugger) [inline src/transformers/feature_extraction_common_utils.py:216]:
```suggestion
                  :func:`~transformers.FeatureExtractionSavingUtilsMixin.save_pretrained` method, e.g.,
```
or adapt to the new name.

File: src/transformers/feature_extraction_common_utils.py [hunk 9]
```diff
@@ -279,12 +252,12 @@ def from_pretrained(
 
 
         Returns:
-            :class:`~transformers.PreTrainedFeatureExtractor`: The feature extractor object instantiated from this
+            :class:`~transformers.SequenceFeatureExtractor`: The feature extractor object instantiated from this
```
Comment (sgugger) [inline src/transformers/feature_extraction_common_utils.py:255]:
```suggestion
            :class:`~transformers.FeatureExtractionSavingUtilsMixin`: The feature extractor object instantiated from this
```

File: src/transformers/feature_extraction_common_utils.py [hunk 10]
```diff
@@ -279,12 +252,12 @@ def from_pretrained(
...
             pretrained model.
 
         Examples::
 
-            # We can't instantiate directly the base class `PreTrainedFeatureExtractor` so let's show the examples on a
+            # We can't instantiate directly the base class `SequenceFeatureExtractor` so let's show the examples on a
```
Comment (sgugger) [inline src/transformers/feature_extraction_common_utils.py:260]:
```suggestion
            # We can't instantiate directly the base class `FeatureExtractionSavingUtilsMixin` so let's show the examples on a
```

File: src/transformers/feature_extraction_common_utils.py [hunk 11]
```diff
@@ -304,7 +276,7 @@ def from_pretrained(
     def save_pretrained(self, save_directory: Union[str, os.PathLike]):
         """
         Save a feature_extractor object to the directory ``save_directory``, so that it can be re-loaded using the
-        :func:`~transformers.PreTrainedFeatureExtractor.from_pretrained` class method.
+        :func:`~transformers.SequenceFeatureExtractor.from_pretrained` class method.
```
Comment (sgugger) [inline src/transformers/feature_extraction_common_utils.py:279]:
```suggestion
        :func:`~transformers.FeatureExtractionSavingUtilsMixin.from_pretrained` class method.
```

File: src/transformers/feature_extraction_common_utils.py [hunk 12]
```diff
@@ -325,7 +297,7 @@ def get_feature_extractor_dict(
     ) -> Tuple[Dict[str, Any], Dict[str, Any]]:
         """
         From a ``pretrained_model_name_or_path``, resolve to a dictionary of parameters, to be used for instantiating a
-        :class:`~transformers.PreTrainedFeatureExtractor` using ``from_dict``.
+        :class:`~transformers.SequenceFeatureExtractor` using ``from_dict``.
```
Comment (sgugger) [inline src/transformers/feature_extraction_common_utils.py:300]:
```suggestion
        :class:`~transformers.FeatureExtractionSavingUtilsMixin` using ``from_dict``.
```

File: src/transformers/feature_extraction_common_utils.py [hunk 13]
```diff
@@ -400,20 +372,20 @@ def get_feature_extractor_dict(
         return feature_extractor_dict, kwargs
 
     @classmethod
-    def from_dict(cls, feature_extractor_dict: Dict[str, Any], **kwargs) -> "PreTrainedFeatureExtractor":
+    def from_dict(cls, feature_extractor_dict: Dict[str, Any], **kwargs) -> PreTrainedFeatureExtractor:
         """
-        Instantiates a :class:`~transformers.PreTrainedFeatureExtractor` from a Python dictionary of parameters.
+        Instantiates a :class:`~transformers.SequenceFeatureExtractor` from a Python dictionary of parameters.
```
Comment (sgugger) [inline src/transformers/feature_extraction_common_utils.py:377]:
```suggestion
        Instantiates a :class:`~transformers.FeatureExtractionSavingUtilsMixin` from a Python dictionary of parameters.
```

File: src/transformers/feature_extraction_common_utils.py [hunk 14]
```diff
@@ -400,20 +372,20 @@ def get_feature_extractor_dict(
...
 
         Args:
             feature_extractor_dict (:obj:`Dict[str, Any]`):
                 Dictionary that will be used to instantiate the feature extractor object. Such a dictionary can be
                 retrieved from a pretrained checkpoint by leveraging the
-                :func:`~transformers.PreTrainedFeatureExtractor.to_dict` method.
+                :func:`~transformers.SequenceFeatureExtractor.to_dict` method.
```
Comment (sgugger) [inline src/transformers/feature_extraction_common_utils.py:383]:
```suggestion
                :func:`~transformers.FeatureExtractionSavingUtilsMixin.to_dict` method.
```

File: src/transformers/feature_extraction_common_utils.py [hunk 15]
```diff
@@ -400,20 +372,20 @@ def get_feature_extractor_dict(
...
             kwargs (:obj:`Dict[str, Any]`):
                 Additional parameters from which to initialize the feature extractor object.
 
         Returns:
-            :class:`~transformers.PreTrainedFeatureExtractor`: The feature extractor object instantiated from those
+            :class:`~transformers.SequenceFeatureExtractor`: The feature extractor object instantiated from those
```
Comment (sgugger) [inline src/transformers/feature_extraction_common_utils.py:388]:
```suggestion
            :class:`~transformers.FeatureExtractionSavingUtilsMixin`: The feature extractor object instantiated from those
```

File: src/transformers/feature_extraction_common_utils.py [hunk 16]
```diff
@@ -447,18 +419,17 @@ def to_dict(self) -> Dict[str, Any]:
         return output
 
     @classmethod
-    def from_json_file(cls, json_file: Union[str, os.PathLike]) -> "PreTrainedFeatureExtractor":
+    def from_json_file(cls, json_file: Union[str, os.PathLike]) -> PreTrainedFeatureExtractor:
         """
-        Instantiates a :class:`~transformers.PreTrainedFeatureExtractor` from the path to a JSON file of parameters.
+        Instantiates a :class:`~transformers.SequenceFeatureExtractor` from the path to a JSON file of parameters.
```
Comment (sgugger) [inline src/transformers/feature_extraction_common_utils.py:424]:
```suggestion
        Instantiates a :class:`~transformers.FeatureExtractionSavingUtilsMixin` from the path to a JSON file of parameters.
```

File: src/transformers/feature_extraction_common_utils.py [hunk 17]
```diff
@@ -447,18 +419,17 @@ def to_dict(self) -> Dict[str, Any]:
...
 
         Args:
             json_file (:obj:`str` or :obj:`os.PathLike`):
                 Path to the JSON file containing the parameters.
 
         Returns:
-            :class:`~transformers.PreTrainedFeatureExtractor`: The feature_extractor object instantiated from that JSON
+            :class:`~transformers.SequenceFeatureExtractor`: The feature_extractor object instantiated from that JSON
```
Comment (sgugger) [inline src/transformers/feature_extraction_common_utils.py:431]:
```suggestion
            :class:`~transformers.FeatureExtractionSavingUtilsMixin`: The feature_extractor object instantiated from that JSON
```

File: src/transformers/feature_extraction_common_utils.py (hunk 5 above)
Comment (patrickvonplaten) [reply src/transformers/feature_extraction_common_utils.py]: Hmm, I was thinking more that we will take the approach of the `generation_utils.py` docstring-like: https://github.com/huggingface/transformers/blob/917f1045025001bc7dfefc992bd101e02fe4bada/src/transformers/generation_utils.py#L769 so that we'll add `... or` statements. Since it's a mixin and doesn't have a `__init__(...)` method I thought, it has to create one of either `PreTrainedSequenceFeatureExtractor` or `PreTrainedImageFeatureExtractor`. Also, regarding the docstring I think it's nicer if the user can directly click on either `PreTrainedSequenceFeatureExtractor` or `PreTrainedImageFeatureExtractor` since the important functionalities are to be found there...

But not sure here...what do you think @LysandreJik @sgugger ?

File: src/transformers/models/wav2vec2/processing_wav2vec2.py [hunk 18]
```diff
@@ -59,7 +59,7 @@ def save_pretrained(self, save_directory):
 
         .. note::
 
-            This class method is simply calling :meth:`~transformers.PreTrainedFeatureExtractor.save_pretrained` and
+            This class method is simply calling :meth:`~transformers.SequenceFeatureExtractor.save_pretrained` and
```
Comment (sgugger) [inline src/transformers/models/wav2vec2/processing_wav2vec2.py:62]:
```suggestion
            This class method is simply calling :meth:`~transformers.FeatureExtractionSavingUtilsMixin.save_pretrained` and
```
sphinx won't resolve the reference since it's not documented on `SequenceFeatureExtractor`. Unless you add the documentation for `SequenceFeatureExtractor` manually in which case you should ignore this suggestion.

It's the same for the other references below.

**Review by sgugger — APPROVED** [review]
Looks good to me, thanks for working on this!

As I said offline, I don't like the very long names for the new classes and the module names so we should strive to find something easier.
